server/app/auth/oidc/OidcProvider.java

```diff
@@ -0,0 +1,185 @@
+package auth.oidc;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import auth.ProfileFactory;
+import com.typesafe.config.Config;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Objects;
+import javax.inject.Provider;
+import org.pac4j.core.http.callback.PathParameterCallbackUrlResolver;
+import org.pac4j.core.profile.creator.ProfileCreator;
+import org.pac4j.oidc.client.OidcClient;
+import org.pac4j.oidc.config.OidcConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import repository.UserRepository;
+
+/**
+ * This class provides the base applicant OIDC implementation. It's abstract because AD and other
+ * providers need slightly different implementations and profile adaptors, and use different config
+ * values.
+ */
+public abstract class OidcProvider implements Provider<OidcClient> {
+
+  private static final Logger logger = LoggerFactory.getLogger(OidcProvider.class);
+  protected final Config configuration;
+  protected final ProfileFactory profileFactory;
+  protected final Provider<UserRepository> applicantRepositoryProvider;
+  protected final String baseUrl;
+
+  protected String[] defaultScopes = {"openid", "profile", "email"};
+
+  private String providerNameConfigName = "provider_name";
```

Comment:
are these constants? make them static final and UPPER_SNAKE_CASE https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

Reply:
+1

should be `private static final` as well